Adding support for "flexible spectral windows"

kartographer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Adds a new optional feature called flexible spectral windows, which adds support for multiple spectral windows, of arbitrary size (in terms of number of channels), channel width, and window direction (i.e., increasing or decreasing frequency with increasing channel number). These changes effectively allow for the spectral window axis (where it exists) within `UVData` to become obsolete.

## Description
<!--- Describe your changes in detail -->
Inspired in part by the way that MIRIAD handles spectral windows, I've added an an optional parameter for `UVData` objects called `flex_spw`, which when set to True (by way of the `_set_flex_spw()` method), allows for one to pack multiple spectral windows of into the frequency axis. Spectral window assignment of individual channels is tracked by the parameter `flex_spw_id_array`, which is an array of shape (Nfreqs,), with values that match to the indices of the `spw_array`. `_set_flex_spw()` also forces the `channel_width` attribute to become an array of shape (Nfreqs,), type=float (originally just type=float). 

Using this new feature, I've added support for Nspws > 1 in MIR (read only), MIRIAD, UVFITS, and UVH5 file formats. I've separately tested to verify that the UVFITS files are being written correctly by loading it into CASA, and have added an additional test data set (from CARMA) which has multiple spectral windows, in order to verify that the above is correctly implemented for MIRIAD. The MIR test data set is already multi-spw, and therefore no changes there were needed. I have forgone adding support for flexible/multi-spw in MS, under the expectation that the relevant module is likely to undergo a significant redesign in the near future anyways.

These changes will effectively make the spectral window axis obsolete, although since removing that axis would change the shape of multiple arrays (which itself would likely be considered a breaking change), I have simply enforced that this axis should always be of length 1 in `UVData` objects (the current effective default, since multi-spw was not previously supported), and added several TODO comments where arrays may need to be reshaped.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The primary motivation for these changes is to allow for support of multiple spectral windows with as few limitations as possible. Many observatories (including SMA, CARMA, ALMA) have the capability of recording data with differing spectral resolution, which is sometimes used to support so-called "line + continuum" observations, where one wants coarse resolution for all but one or two spectral windows, which presumably have some relatively narrow spectral line of interest. These changes also provide better support for irregularly sampled spectral windows, which may have gaps due to some combination of flagging and frequency averaging. 

<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
This PR addresses the following issues:
- Closes #1 
- Closes #759 

After merge, I am tentatively planning on filing the following as follow-up issues (they could in principle be added to this PR, if desired):
- Adding support for flexible spws with the `frequency_average` method of `UVData`
- Adding support for flexible spws with `UVBeam` objects
- Adding support for flexible spws with `UVCal` objects
- Adding support for flexible spws with `UVFlag` objects
- Remove the vestigial spw-axis

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

bhazelton commented on pyuvdata/uvdata/uvfits.py:
```diff
@@ -437,38 +446,62 @@ def read_uvfits(
+                fq_hdu = hdu_list[hdunames["AIPS FQ"]]
+                assert self.Nspws == fq_hdu.header["NO_IF"]
+
+                # TODO: This is fine for now, although I (karto) think that this
+                # is relative to the rest_freq, which can be specified as part of
+                # the AIPS SU table.
```
I just spent some time reading the uvfits memo and found this:

>The AIPS FQ or frequency table is an essential table unless there is only one spectral window and one frequency setup. It conveys the frequency offsets to be added to the reference frequency as a function of spectral window and frequency setup. The actual observing frequency requires the addition of source-dependent frequency offsets from the AIPS SU table (Section 4.17 on page 34) and time-dependent frequency offsets from the AIPS FO table (Section 4.6 on page 21).

So my read is that the value in the visibility header is the "reference frequency" rather than the "rest frequency" and that if there is an SU table you also need to add in the offsets contained in that table. Does that seem right?

kartographer replied:
It does -- this should have been written as `ref_frequency` (which is actually what Memo 117 calls it). I put in a note about rest frequencies in anticipation of some additional future changes (similar to mentions of SOURCE, in anticipation of multi-source datasets), but then the phrase "rest frequency" clearly got stuck in my brain. Fixed, pushed in 1be5163

bhazelton commented:
I think it's fine to make all those next steps issues in their own right rather than doing it all in this PR. But I'd really like to get the changes into the other objects as soon as we can after we merge this one.

kartographer commented on pyuvdata/uvdata/miriad.py:
```diff
@@ -357,7 +368,15 @@ def read_miriad(
+                        warn_extra_sources = False
+                    continue
+                raise NotImplementedError(
                     "This appears to be a multi source file, which is not supported."
```
Good idea -- added it, pushed in 1be5163

kartographer commented on pyuvdata/uvdata/miriad.py:
```diff
@@ -786,9 +810,10 @@ def write_miriad(
-        if self.data_array.shape[1] > 1:
-            raise ValueError("write_miriad currently only handles single spw files.")
+        # if self.data_array.shape[1] > 1:
+        #    raise ValueError("write_miriad currently only handles single spw files.")
```
Nope -- deleted, pushed in 1be5163

kartographer commented on pyuvdata/uvdata/tests/test_mir.py:
```diff
@@ -122,19 +124,81 @@ def test_write_mir(uv_in_uvfits, err_type=NotImplementedError):
     """
     testfile = os.path.join(DATA_PATH, "sma_test.mir")
     uv_in = UVData()
-    with pytest.raises(err_type, match=err_msg):
```
That's fair -- I've gone ahead and added back the matching text. Pushed in 1be5163

kartographer commented on pyuvdata/uvdata/tests/test_miriad.py:
```diff
@@ -108,6 +108,74 @@ def test_read_nrao_write_miriad_read_miriad(casa_uvfits, tmp_path):
+    uv_in = UVData()
+    uv_out = UVData()
+    carma_file = os.path.join(DATA_PATH, "carma_miriad")
+    testfile = str(tmp_path / "outtest_carma_miriad.uv")
```
Done (both here and in about a dozen other places in `test_miriad`), pushed 1be5163

kartographer commented on pyuvdata/uvdata/uvfits.py:
```diff
@@ -848,6 +938,12 @@ def write_uvfits(
 
+        # TODO Karto: Mark why here
+        # if self.Nspws > 1:
+        #     parnames_use += ["FREQSEL "]
```
No -- it was a leftover from early testing with UVFITS. Deleted, pushed in 1be5163

kartographer commented on pyuvdata/uvdata/uvfits.py:
```diff
@@ -1043,10 +1141,40 @@ def write_uvfits(
+        # If needed, add the FQ table
+        if self.Nspws > 1:
+            # skipping for now and limiting to a single spw
+            # Karto: SKIP NO MORE!
```
Fair enough =) Pushed in 1be5163

kartographer commented on pyuvdata/uvdata/uvh5.py:
```diff
@@ -729,7 +739,9 @@ def _get_data(
-                assert self.Nspws == visdata.shape[1]
+                # Commenting this out for now, since we are explicitly breaking this
+                # association between data shape and Nspws
+                # assert self.Nspws == visdata.shape[1]
```
Done. Pushed in 1be5163

kartographer commented on pyuvdata/uvdata/tests/test_miriad.py:
```diff
@@ -108,6 +108,74 @@ def test_read_nrao_write_miriad_read_miriad(casa_uvfits, tmp_path):
+    uv_in.phase_center_ra = 0.0
+    uv_in.phase_center_dec = 0.0
+    uv_in.phase_center_epoch = 2000.0
+    uv_in.extra_keywords = {}  # TODO: Capture these extra keywords
```
Most of the extra keywords are of the above forbidden types, but I've made a more explicit check to look for things that are dicts, lists, or arrays (and if so, then pop then from the `extra_keywords` dict). Pushed in 1be5163

kartographer commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -287,6 +292,34 @@ def __init__(self):
+            description=desc,
+            expected_type=np.bool,
+            value=False,
+            required=False,
```
Okay, that makes sense -- I was originally thinking that it was required, but then started waffling back and forth on it. Fixed, pushed in 1be5163

kartographer commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -778,40 +825,117 @@ def set_lsts_from_time_array(self, background=False):
-                    "frequencies that are spaced by more than their "
-                    "channel width."
-                )
+                print(freq_spacing[0, 0], self.channel_width, self._channel_width.tols)
```
Whoops -- I guess you can see what the last thing was I was attempting to debug. Deleted, pushed in 1be5163

kartographer commented on pyuvdata/uvdata/uvfits.py:
```diff
@@ -719,12 +752,53 @@ def write_uvfits(
+                )
+
+            # Make sure freq values are greater zero. Note that I think _technically
+            # one could write negative frequencies into the dataset, but I am pretty
+            # sure that reduction packages may balk hard.
```
I originally wrote "barf hard", but then thought maybe that a little too graphic ;)

kartographer commented on pyuvdata/uvdata/mwa_corr_fits.py:
```diff
@@ -116,7 +116,7 @@ def flag_init(
             )
 
-        num_ch_flag = int(edge_width / self.channel_width)
+        num_ch_flag = int(edge_width / np.abs(self.channel_width))
```
So as written, `channel_width` encodes both the width of the channel and which way to spectral window goes. As originally implemented, MIRIAD (which encodes this as `sdf`), UVFITS (`CDELT4` or `CH WIDTH`), and MIR (`fres`) all allow this value to be negative, to signify that the window goes from higher to lower frequencies with increasing channel number. I came across this when I was doing bulk searches for the `channel_width` and `freq_array` attributes, and noticed that he above was incompatible with storing these values as negative, so I put in the small fix above. I could alternatively change the code to force `channel_width` to always be positive (which would only require a few additional lines of code).

bhazelton replied:
ah, interesting. Yes, I suppose FITS does in principal support that. I think we want the `channel_width` parameter on the object to be positive definite.

kartographer replied:
Okay, after the world's least interesting game of whack-a-mole, `channel_width` is now > 0. Pushed in 8853b902

kartographer commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -241,11 +243,14 @@ def __init__(self):
+            "Width of frequency channels (Hz). If flex_spw = False, then it single "
+            "value of type = float, otherwise it is an array of shape (Nfreqs), type = "
+            "float."
+        )
         self._channel_width = uvp.UVParameter(
-            "channel_width",
-            description="Width of frequency channels (Hz)",
-            expected_type=np.float,
-            tols=1e-3,
+            "channel_width", description=desc, expected_type=np.float, tols=1e-3,
         )  # 1 mHz
```
I'm certainly happy to just force this to be an array -- while implementing the changes, I was attempting to minimize the potential for any breaking changes, which is why I allowed it to remain just a solitary float. The only place that one would need to take care to wrap this would be w/ older UDH5 files, because of course this is effectively encoding a change to the file structure. I do already have a catch build in for `read_uvh5`, so it wouldn't be difficult to expand that just a tad (all that to say: "I'm happy to go with whatever makes you guys happy").

bhazelton replied:
I've scheduled a conversation with HERA folks, I'll get back to you after that about this. Your approach is safer, I agree, so we'll probably stick with that for this PR.

kartographer commented on pyuvdata/uvdata/uvh5.py:
```diff
@@ -524,8 +530,12 @@ def _read_header(
-        self.channel_width = float(header["channel_width"][()])
         self.spw_array = header["spw_array"][:]
+        # We've added a new keyword, that did not exist before, so
+        if self.flex_spw:
+            self.channel_width = header["channel_width"][:]
+        else:
+            self.channel_width = float(header["channel_width"][()])
```
Yup -- we'd still need the split handling if `flex_spw` isn't found as a keyword in the header (since that implies that `channel_width` is a solitary float), so some version of this would get moved up into line 448.

kartographer commented:
Thanks @bhazelton -- with respect to the additional issues, I agree that we'd like to get them settled sooner rather than later (definitely before the next release). @ChrisMoriarty were chatting a bit, and I think we'll attempt to do a bit of divide-and-conquer on this front. Based on our discussion last Wednesday, it seems like there is a different point person within the pyuvdata management structure for us to confer on changes for the UVFlag, UVCal, and UVBeam objects, and so it seemed sensible to craft a new branch/PR for each one.

kartographer commented on pyuvdata/uvdata/uvfits.py:
```diff
@@ -1032,7 +1129,8 @@ def write_uvfits(
         # -- lists of spws given in a SU table.
-        ant_hdu.header["FREQID"] = -1
+        # Karto: Here might be a place to address freq setup?
+        ant_hdu.header["FREQID"] = 1
```
Yeah, so this is a byproduct of the "axes all the way down" cosmology of UVFITS -- you can, in theory, have multiple "instrument setups", so long as they all have the same number of channels, windows, baseline-times, polarizations, etc. 99.9% of the time, I think if you have all those things matching, it's probably just more sensible to concatenate the data along the spw axis (the SMA has one use case where this isn't true, that's easier to talk through than type). In practice, this value is just a dummy value, but it is required in the frequency table headers, so I changed this to match the value below.

bhazelton replied:
ok, thanks for that explanation!

bhazelton commented on pyuvdata/uvdata/miriad.py:
```diff
@@ -328,22 +333,28 @@ def read_miriad(
-                    "window (spw) are not yet supported. A great "
-                    "project for the interested student!"
-                )
+                # self.Nspws = d.shape[0]
+                # self.spw_array = np.arange(self.Nspws)
```
should we just remove these lines?

kartographer replied:
Agreed -- deleted, pushed in 8f54839.

bhazelton commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -241,11 +243,14 @@ def __init__(self):
         )  # 1 ms
+
+        desc = (
+            "Width of frequency channels (Hz). If flex_spw = False, then it single "
+            "value of type = float, otherwise it is an array of shape (Nfreqs), type = "
```
"it single" -> "it is a single"

kartographer replied:
Fixed, pushed in 436bc6c.

kartographer commented on pyuvdata/uvdata/uvh5.py:
```diff
@@ -445,6 +445,12 @@ def _read_header(
             ).decode("utf8")
 
+        if "flex_spw" in header:
+            if bool(header["flex_spw"][()]):
```
So on the SMA front, there are a few additional things that we'd like to add that may have an impact on the UVH5 file format -- the example I can most readily point to would be the ability to have multiple sources in 'phased' mode (following a similar model to what we used for flexible spectral windows), but there are a few others that I can think of. Maybe we can broaden the scope of the issue to cover those, and I can also enumerate said expansions now, and/or we can discuss it further on said issue.

plaplant replied:
Thanks for the info. I'd vote for making an issue(s) about the changes we'd like to make to the object, and then I'll make the UVH5 memo issue broader to reference the resulting changes to the file format that will support them.

kartographer commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -287,6 +292,31 @@ def __init__(self):
+        desc = (
+            "Required if flex_spw = True. Maps individual channels along the "
+            "frequency axis to individual spectral windows, as listed in the "
+            "spw_array (zero-indexed). Shape (Nspws), type = int."
```
Whoops -- good catch, fixed, pushed in e99c790.

kartographer commented on pyuvdata/uvdata/uvh5.py:
```diff
@@ -524,8 +530,12 @@ def _read_header(
         self.freq_array = header["freq_array"][:, :]
-        self.channel_width = float(header["channel_width"][()])
         self.spw_array = header["spw_array"][:]
+        # We've added a new keyword, that did not exist before, so
```
That was actually a comment that got left behind (and cut off, apparently) when I reorganized the code a little bit. I've moved the comment up to line 448, where its highlighting the fact that we need an extra bit of wrapping, because even though it's a required parameter, older versions of the UVH5 file format won't have this keyword set. Comments updated, pushed in 89a06d3.

kartographer commented:
Thanks @bhazelton and @plaplant! I think I've addressed everything above, and have gone ahead and updated the README and the relevant section of the tutorials (and checked to make sure that the CHANGELOG looks good). Hopefully that just about covers it, but let me know if anything else is needed.

bhazelton reviewed (approved):
It looks good to me. Are you happy @plaplant ?

bhazelton commented:
Oops, it looks like this branch needs to be rebased before it can be merged in.

kartographer commented:
Always just one more thing to do... ;) I'll pull over from master and sort out the merge.

kartographer commented:
Okay @bhazelton, @plaplant -- all patched up again, with a corrected CHANGELOG to boot.

plaplant reviewed (approved):
Thanks @kartographer! Looks great. Fine by me to merge anytime.

kartographer commented:
Well that's embarrassing -- looks like I didn't do the rebase correctly. I've tried rebasing one more time and found the conflict, hopefully it sticks this time (and if not, I may call in the cavalry).

kartographer commented:
And FWIW -- the conflict was solely in the CHANGELOG, nothing else got touched on rebase.

plaplant reviewed (approved):
That's what we get for trying to be responsible and have an accurate CHANGELOG 😉
